### ISSUE TITLE
Phase 4: Dreaming Cycle - Autonomous Prompt Optimization (Issue #27)

### ISSUE THREAD

**Issue by chriskevini** [chriskevini/lattice#34]
## Summary

Implement the Dreaming Cycle: autonomous prompt optimization system that analyzes audit data, proposes improvements, and enables human-approved self-evolution. This is the culmination of Issue #27 - the system becomes self-improving.

## Background

Phases 1-3 established:
- Phase 1: Audit schema for tracking (COMPLETED)
- Phase 2: Complete audit system with feedback (COMPLETED)
- Phase 3: Analysis queries and pattern recognition (Issue #33)

Phase 4 completes the loop: **The system analyzes itself and proposes improvements.**

## The Dreaming Cycle Concept

```
┌─────────────────────────────────────────────────────────────────┐
│                      THE DREAMING CYCLE                         │
├─────────────────────────────────────────────────────────────────┤
│                                                                 │
│   1. ANALYZE                                                   │
│      ↓ Audit data + feedback + effectiveness                    │
│   2. IDENTIFY                                                   │
│      ↓ Underperforming prompts + patterns                       │
│   3. PROPOSE                                                    │
│      ↓ Generate optimization suggestions                        │
│   4. VALIDATE                                                   │
│      ↓ Test suggestions against constraints                     │
│   5. PROPOSE CHANGE                                             │
│      ↓ Create dream channel proposal with before/after          │
│   6. HUMAN APPROVE                                              │
│      ↓ Human reviews in dream channel (emoji reaction)          │
│   7. APPLY                                                      │
│      ↓ Update prompt_registry if approved                       │
│   8. MEASURE                                                    │
│      ↓ Track improvement in Phase 3 metrics                     │
│                                                                 │
└─────────────────────────────────────────────────────────────────┘
```

### 1. Analysis Engine (`lattice/dreaming/analyzer.py`)

```python
class DreamingAnalyzer:
    async def identify_optimization_targets(self) -> list[OptimizationTarget]:
        \"\"\"Find prompts that need optimization.\"\"\"
        # Queries Phase 3 analysis results
        # Prioritizes by:
        # - High usage + low effectiveness = high priority
        # - Recurring negative feedback = high priority
        # - Performance (cost/latency) issues = medium priority
        
class OptimizationTarget:
    prompt_key: str
    current_metrics: PromptEffectivenessReport
    issues: list[str]  # e.g., ["high latency", "negative feedback pattern"]
    priority: int  # 1-5
```

### 2. Optimization Generator (`lattice/dreaming/generator.py`)

```python
class PromptOptimizer:
    async def generate_optimization(
        self,
        target: OptimizationTarget,
    ) -> PromptOptimizationProposal:
        \"\"\"Generate optimization suggestion for a prompt.\"\"\"
        # Analyzes current prompt template
        # Identifies specific improvements:
        # - Add/remove context sections
        # - Adjust archetype configuration
        # - Modify instructions
        # - Change temperature/parameters
        
class PromptOptimizationProposal:
    prompt_key: str
    current_template: str
    proposed_template: str
    expected_improvements: list[str]  # e.g., ["reduced latency by 20%", "better context usage"]
    confidence: float  # 0.0-1.0
    rationale: str
```

### 3. Proposal System (`lattice/dreaming/proposals.py`)

```python
class ProposalManager:
    async def create_proposal(
        self,
        optimization: PromptOptimizationProposal,
    ) -> UUID:
        \"\"\"Create dream channel proposal for human review.\"\"\"
        # Stores proposal in database
        # Creates rich embed in dream channel
        # Returns proposal ID
        
    async def handle_approval(
        self,
        proposal_id: UUID,
        approved: bool,
        human_feedback: str | None = None,
    ) -> bool:
        \"\"\"Process human approval/rejection.\"\"\"
        # If approved: apply optimization to prompt_registry
        # If rejected: log feedback, mark proposal closed
        # Return success status
```

### 4. Dream Channel Proposal Format

**Automated Dreaming Cycle Proposal:**

```
🌙 **DREAMING CYCLE: PROPOSED OPTIMIZATION**

**Target Prompt:** BASIC_RESPONSE
**Priority:** 🔴 HIGH (high usage, negative feedback patterns)

📊 **CURRENT STATE:**
- Usage: 247 times
- Feedback Rate: 12% (below target 15%)
- Avg Latency: 340ms (high)
- Common Feedback: \"too slow\", \"unclear response\"

📈 **PROPOSED CHANGES:**

1. **Reduce Context**
   Current: 10E, 10S, 5G
   Proposed: 5E, 5S, 2G
   Expected: -30% latency

2. **Simplify Instructions**
   Current: 150 tokens of instructions
   Proposed: 100 tokens (more direct)
   Expected: Better focus, faster generation

📝 **PROPOSED TEMPLATE (diff):**
```diff
- SYSTEM: You are a helpful Discord companion.
+ SYSTEM: You are a concise, helpful Discord companion.

- INSTRUCTIONS: 
- 1. Read recent context carefully
- 2. Consider semantic memory
- 3. Check graph relationships
- 4. Formulate response
+ INSTRUCTIONS: 
+ 1. Check recent messages
+ 2. Answer directly
```

🤖 **CONFIDENCE:** 78%

**This is an automated Dreaming Cycle proposal.**
React with ✅ to approve, ❌ to reject.

---

**After Human Approval:**

```
🌙 **DREAMING CYCLE: APPROVED** ✅

**BASIC_RESPONSE** has been optimized.

**Changes Applied:**
- Context reduced (10E→5E, 10S→5S, 5G→2G)
- Instructions simplified (150→100 tokens)

**Expected Impact:**
- -30% latency
- Better response focus

**Improvement will be measured in Phase 3 analysis.**
```

### 5. Approval Workflow

**Emoji Reactions:**
- ✅ = Approve and apply
- ❌ = Reject (requires reason)
- 🤔 = Need more analysis (triggers deeper investigation)
- ⏸️ = Defer for later review

**Approval Rules:**
- Must be in dream channel
- Must be bot's own message
- Must have ✅ reaction from human with permissions

### 6. Version Control for Prompts

```sql
-- Track prompt changes over time
CREATE TABLE prompt_versions (
    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
    prompt_key TEXT NOT NULL,
    version INTEGER NOT NULL,
    template TEXT NOT NULL,
    context_config JSONB,
    created_by TEXT NOT NULL,  -- 'human' or 'dreaming_cycle'
    proposal_id UUID REFERENCES dreaming_proposals(id),
    created_at TIMESTAMPTZ DEFAULT now(),
    UNIQUE(prompt_key, version)
);

-- Dreaming proposals table
CREATE TABLE dreaming_proposals (
    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
    prompt_key TEXT NOT NULL,
    current_version INTEGER NOT NULL,
    proposed_version INTEGER NOT NULL,
    current_template TEXT NOT NULL,
    proposed_template TEXT NOT NULL,
    rationale TEXT NOT NULL,
    expected_improvements JSONB,
    confidence FLOAT,
    status TEXT DEFAULT 'pending' CHECK (status IN ('pending', 'approved', 'rejected', 'deferred')),
    human_feedback TEXT,
    created_at TIMESTAMPTZ DEFAULT now(),
    reviewed_at TIMESTAMPTZ,
    reviewed_by TEXT
);
```

### 7. Scheduler Integration

Run dreaming cycle periodically (e.g., daily):

```python
# In scheduler/triggers.py
async def run_dreaming_cycle() -> None:
    \"\"\"Execute one iteration of the dreaming cycle.\"\"\"
    analyzer = DreamingAnalyzer()
    optimizer = PromptOptimizer()
    proposals = ProposalManager()
    
    # 1. Identify optimization targets
    targets = await analyzer.identify_optimization_targets()
    
    # 2. Generate proposals for high-priority targets
    for target in targets[:3]:  # Limit to 3 per run
        if target.priority >= 4:  # Only HIGH priority
            optimization = await optimizer.generate_optimization(target)
            await proposals.create_proposal(optimization)
    
    # 3. Log results
    logger.info(
        "Dreaming cycle complete",
        targets_identified=len(targets),
        proposals_created=len(targets),
    )
```

## Dependencies

**Required (Must implement first):**
- Issue #27 Phase 2 - Prompt audit system with feedback (COMPLETED)
- Issue #33 (Phase 3) - Analysis queries and pattern recognition (MUST COMPLETE FIRST)
- This issue (Phase 4) builds directly on Phase 3 output

**Recommended (Nice to have, not required):**
- Issue #32 - UI components for proposal display (nice to have but not required)
- Issue #XX - Prompt template CRUD API (needed for production management)

**Technical Dependencies:**
- Phase 3 analysis results
- Prompt registry table (already exists)
- Dream channel access (already exists)

## Files to Create

- `lattice/dreaming/__init__.py` (new)
- `lattice/dreaming/analyzer.py` (new)
- `lattice/dreaming/generator.py` (new)
- `lattice/dreaming/proposals.py` (new)
- `lattice/dreaming/versioning.py` (new)
- `scripts/migrations/006_add_dreaming_schema.sql` (new)

## Files to Modify

- `lattice/scheduler/triggers.py` - Add dreaming_cycle trigger
- `lattice/discord_client/bot.py` - Add approval reaction handler

## Testing

- Unit tests for analyzer, generator, proposals
- Integration tests with audit data
- Approval workflow tests (emoji handling)
- Version control tests (rollback capability)
- Scheduler integration tests

## Success Criteria

- [ ] Dreaming cycle runs automatically on schedule
- [ ] Proposals created for underperforming prompts
- [ ] Human approval workflow works in dream channel
- [ ] Prompts are versioned when modified
- [ ] Approved changes are applied to prompt_registry
- [ ] Improvement is measurable in Phase 3 analysis
- [ ] Tests pass (target: 80% coverage on dreaming module)

## Safety Measures

1. **Confidence Threshold**: Only propose if confidence > 0.7
2. **Version Control**: All changes tracked, can rollback
3. **Human Approval**: All changes require human approval
4. **Rate Limiting**: Max 3 proposals per cycle to avoid spam
5. **Circuit Breaker**: Pause dreaming if too many rejections

## Example Flow

```
Day 1: Dreaming cycle runs
→ Identifies BASIC_RESPONSE as underperforming
→ Generates optimization proposal
→ Posts to dream channel

Day 2: Human reviews proposal
→ React with ✅ (approve)
→ Change is applied
→ New version created (v1 → v2)

Day 7: Dreaming cycle runs again
→ Compares v2 metrics to v1
→ Reports improvement in dream channel
→ Or generates further optimization if needed
```

## Notes

- Start with conservative confidence threshold (0.8+)
- Log all proposals even rejected ones (training data)
- Consider A/B testing for major changes
- Allow manual trigger: `!dream` command in dream channel
- Dreaming cycle should run during low-activity hours

## Follow-up

After Phase 4:
- Monitor improvement rates
- Refine confidence scoring
- Add automated A/B testing
- Consider full autonomy for minor tweaks
- Build feedback loop into prompt generation

Closes #27 (Phase 4 - COMPLETION)
